### PR TITLE
ENH: Add nullable keyword to read_sql

### PR DESCRIPTION
**Pull request by phofl**
Sits on top of #50047 

Functionality wise, this should work now.

More broadly, I am not sure that this is the best approach we could take here. Since the ``convert_to_nullable_type`` in ``lib.maybe_convert_objects`` is not used right now except here, we could also make this strict and return the appropriate Array from the Cython code, not only when nulls are present. This would avoid the re-cast in the non-cython code part.

### CODE REVIEW

**Comment by WillAyd** [comment]
I think makes sense and like your vision on the interaction with Cython. Is there any performance difference here moving away from the `.from_records` construction of the DataFrame?

**Comment by phofl** [comment]
Also discussed this with @mroeschke in another pr, he leaned more towards keeping maybe_convert_objects as is.

Maybe a bit faster? The previous implementation was running through the same function that we are calling directly now. Definitively no slowdown

**Review by WillAyd — APPROVED** [review]
I'm OK with this as is - @mroeschke

**Comment by mroeschke** [comment]
>  Since the convert_to_nullable_type in lib.maybe_convert_objects is not used right now except here, we could also make this strict and return the appropriate Array from the Cython code, not only when nulls are present

I think in a future PR it might make sense to do this; I didn't realized this was the only location where this is used. I guess the only slight downside is run time importing the Arrays in the Cython code?

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

File: pandas/core/internals/construction.py [hunk 1]
```diff
@@ -973,20 +984,37 @@ def _convert_object_array(
...
+                try_float=coerce_float,
+                convert_to_nullable_dtype=use_nullable_dtypes,
+            )
 
             if dtype is None:
                 if arr.dtype == np.dtype("O"):
                     # i.e. maybe_convert_objects didn't convert
                     arr = maybe_infer_to_datetimelike(arr)
+                    if use_nullable_dtypes and arr.dtype == np.dtype("O"):
+                        arr = StringDtype().construct_array_type()._from_sequence(arr)
```
Comment (jbrockmendel) [inline pandas/core/internals/construction.py:1009]: this seems weird to me.  why are we casting potentially non-strings to strings?

File: pandas/core/internals/construction.py [hunk 2]
```diff
@@ -973,20 +984,37 @@ def _convert_object_array(
...
+                elif use_nullable_dtypes and isinstance(arr, np.ndarray):
+                    if is_integer_dtype(arr.dtype):
+                        arr = IntegerArray(arr, np.zeros(arr.shape, dtype=np.bool_))
+                    elif is_bool_dtype(arr.dtype):
+                        arr = BooleanArray(arr, np.zeros(arr.shape, dtype=np.bool_))
+                    elif is_float_dtype(arr.dtype):
```
Comment (jbrockmendel) [inline pandas/core/internals/construction.py:1015]: could re-use pd.array for L1011-L1016?

File: pandas/io/sql.py [hunk 3]
```diff
@@ -139,24 +140,46 @@ def _parse_date_columns(data_frame, parse_dates):
...
+    coerce_float: bool = True,
+    use_nullable_dtypes: bool = False,
+) -> DataFrame:
+    content = lib.to_object_array_tuples(data)
+    arrays = convert_object_array(
+        list(content.T),
+        dtype=None,
+        coerce_float=coerce_float,
+        use_nullable_dtypes=use_nullable_dtypes,
+    )
```
Comment (jbrockmendel) [inline pandas/io/sql.py:155]: looks like this has gotten some Arrow-specific logic added here in the interim.  can we move the backend-specific stuff from core.internals here?

**Pull request merged**